Framework/Core/include/Framework/TableTreeHelpers.h

```diff
@@ -22,97 +22,205 @@ namespace o2::framework
+    } else {
+      mLeaflist = mBranchName + ROOTTypeString_t<T>;
+    }
+    mBranch = tree->GetBranch(mBranchName.c_str());
+    if (mBranch == nullptr) {
+      mBranch = tree->Branch(mBranchName.c_str(), (char*)nullptr, mLeaflist.c_str());
+    }
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      mCurrent = new bool[listSize];
+      mLast = mCurrent + listSize;
+      accessChunk(0);
+    } else {
+      accessChunk(0);
+    }
+  }
 
-  // add branches
-  bool addBranch(std::shared_ptr<arrow::ChunkedArray> col, std::shared_ptr<arrow::Field> field);
-  bool addAllBranches();
+ private:
+  std::string mLeaflist;
+
+  void resetBuffer() override
+  {
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      if (O2_BUILTIN_UNLIKELY((*mCurrentPos - mFirstIndex) * listSize >= getCurrentArray()->length())) {
+        nextChunk();
+      }
+    } else {
+      if (O2_BUILTIN_UNLIKELY((mCurrent + (*mCurrentPos - mFirstIndex) * listSize) >= mLast)) {
+        nextChunk();
+      }
+    }
+    accessChunk(*mCurrentPos);
+    mBranch->SetAddress((void*)(mCurrent + (*mCurrentPos - mFirstIndex) * listSize));
+  }
+
+  auto getCurrentArray() const
+  {
+    if (listSize > 1) {
+      return std::static_pointer_cast<o2::soa::arrow_array_for_t<std::remove_pointer_t<T>>>(std::static_pointer_cast<arrow::FixedSizeListArray>(mColumn->chunk(mCurrentChunk))->values());
+    } else {
+      return std::static_pointer_cast<o2::soa::arrow_array_for_t<std::remove_pointer_t<T>>>(mColumn->chunk(mCurrentChunk));
+    }
+  }
+
+  void nextChunk() override
+  {
+    mFirstIndex += getCurrentArray()->length();
+    ++mCurrentChunk;
+  }
+
+  void accessChunk(int at)
+  {
+    auto array = getCurrentArray();
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      for (auto i = 0; i < listSize; ++i) {
+        mCurrent[i] = (bool)array->Value((at - mFirstIndex) * listSize + i);
+      }
+    } else {
+      mCurrent = (std::remove_pointer_t<T>*)array->raw_values();
+      mLast = mCurrent + array->length() * listSize;
+    }
+  }
+
+  mutable std::remove_pointer_t<T>* mCurrent = nullptr;
+  mutable std::remove_pointer_t<T>* mLast = nullptr;
+  TBranch* mBranch = nullptr;
+};
+
+class TableToTree
+{
+ public:
+  TableToTree(std::shared_ptr<arrow::Table> table, TFile* file, const char* treename);
 
-  // write table to tree
   TTree* process();
+  void addBranch(std::shared_ptr<arrow::ChunkedArray> column, std::shared_ptr<arrow::Field> field);
+  void addAllBranches();
+
+ private:
+  arrow::Table* mTable;
+  int64_t mRows = 0;
+  TTree* mTree = nullptr;
+  std::vector<std::unique_ptr<ColumnToBranchBase>> mColumnReaders;
+};
+
+class GenericTreeToTable
+{
+ public:
+ private:
+  std::shared_ptr<arrow::Table> mTable;
+  std::vector<std::shared_ptr<arrow::Field>> mFields;
+  std::string mTableLabel;
 };
```

Comment:
??

Reply:
A leftover from splitting, I'll remove it.

Reply:
Removed.